server/aesdsocket.c

```diff
@@ -0,0 +1,253 @@
+    if (listen(server_socket, 1) == -1) {
+        perror("Listen failed"); // Print an error message if listen fails
+        syslog(LOG_ERR, "Listen failed: %m"); // Log listen error
+        close(server_socket); // Close the server socket
+        return -1; // Return an error status
+    }
+
+    while (1) 
+    {
+        struct sockaddr_in client_addr;
+        socklen_t client_len = sizeof(client_addr);
+
+        client_socket = accept(server_socket, (struct sockaddr *)&client_addr, &client_len);
+        if (client_socket == -1) {
+            perror("Accept failed"); // Print an error message if accept fails
+            syslog(LOG_ERR, "Accept failed: %m"); // Log accept error
+        }
+        else {
+            char client_ip[INET_ADDRSTRLEN];
+
+            getpeername(client_socket, (struct sockaddr *)&client_addr, &client_len);
+
+            inet_ntop(AF_INET, &(client_addr.sin_addr), client_ip, INET_ADDRSTRLEN);
+
+            // Log accepted connection
+            syslog(LOG_INFO, "Accepted connection from %s", client_ip);
+
+            char *buffer = (char *)malloc(sizeof(char) * BUFFER_SIZE);
+            if (buffer == NULL) {
+                syslog(LOG_ERR, "Memory allocation failed!"); // Log memory allocation error
+                return -1; // Return an error status
+            }
+
+            // Receive and append data
+            int data_file = open(DATA_FILE, O_WRONLY | O_CREAT | O_APPEND, FILE_PERMISSIONS);
+            if (data_file == -1) {
+                perror("File open failed"); // Print an error message if file open fails
+                syslog(LOG_ERR, "File open failed: %m"); // Log file open error
+                free(buffer);
+                close(client_socket);
+            }
+            else {
+                ssize_t bytes_received;
+                while ((bytes_received = recv(client_socket, buffer, BUFFER_SIZE, 0)) > 0) {
+                    write(data_file, buffer, bytes_received);
+                    if (memchr(buffer, '\n', bytes_received) != NULL) {
+                        break;
+                    }
+                }
+
+                syslog(LOG_INFO, "Received data from %s: %s, %d", client_ip, buffer, (int)bytes_received);
+
+                close(data_file);
+
+                lseek(data_file, 0, SEEK_SET);
+
+                // Send data back to the client
+                data_file = open(DATA_FILE, O_RDONLY);
+                if (data_file == -1) {
+                    perror("file open failed"); // Print an error message if file open fails
+                    syslog(LOG_ERR, "File open failed: %m"); // Log file open error
+                    free(buffer);
+                    close(client_socket);
+                }
+                else {
+                    ssize_t bytes_read;
+                    memset(buffer, 0, sizeof(char) * BUFFER_SIZE);
+
+                    while ((bytes_read = read(data_file, buffer, sizeof(char) * BUFFER_SIZE)) > 0) {
```

Comment:
What if \n was not the last character? you are still reading the unrequired characters to the data_file fd and sending them via socket

Reply:
Yeah I totally missed it, will add it. Thanks for pointing it out.